x-pack/plugins/ml/public/application/components/model_snapshots/model_snapshots_table.tsx

```diff
@@ -50,15 +50,24 @@ export const ModelSnapshotTable: FC<Props> = ({ job, refreshJobList }) => {
   const [closeJobModalVisible, setCloseJobModalVisible] = useState<ModelSnapshot | null>(null);
   const [combinedJobState, setCombinedJobState] = useState<COMBINED_JOB_STATE | null>(null);
 
+  let unmounted = false;
```

Comment:
To avoid the `let` and resetting the variable on every render we use this alternative pattern in some places: https://github.com/elastic/kibana/blob/master/x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_exploration/components/outlier_exploration/use_outlier_data.ts#L81

We discussed making a reusable custom hook in the past but didn't get around to do it. Here's the PR that originally introduced it: https://github.com/elastic/kibana/pull/81123

Reply:
I had a look at other similar examples in our code, but could not find one where the guard variable is used in a separate function and so needed to in the outer scope.


I might missing something, but using a `const` wrapper around the variable won't make any difference? Both will be redefined on render.
```
let unmounted = false;
```
vs
```
const options = { unmounted: false };
```

Reply:
You're right this is different because of the additional outside callback. In the other examples the difference was that everything could be kept within the `useEffect`.

As an alternative you could use a refresh value to trigger the `useEffect` instead of directly calling `loadModelSnapshots()`.

Something like:

```ts
const [isRefreshed, setIsRefreshed] = useState(true);

useEffect(() => {
    const options = { didCancel: false };
    loadModelSnapshots(options);
    return () => {
      options.didCancel = true;
    };
  }, [isRefreshed]);

async function loadModelSnapshots(options) {
    if (options.didCancel === true) {
      // table refresh can be triggered a while after a snapshot revert has been triggered.
      // ensure the table is still visible before attempted to refresh it.
      return;
    }
    const { model_snapshots: ms } = await ml.getModelSnapshots(job.job_id);
    setSnapshots(ms);
    setSnapshotsLoaded(true);
}

const refresh = useCallback(
    () => {
        setIsRefreshed(state => !state);
        // wait half a second before refreshing the jobs list
        setTimeout(refreshJobList, 500);
    }
    [setIsRefreshed],
);
```

Reply:
I've updated this check so it now uses `useRef` to keep track of the flag for whether the component is mounted.
in 6f1867da0d5ece31d12035d186f9eeda3e18797f